src/main/groovy/com/google/protobuf/gradle/ProtobufPlugin.groovy

```diff
@@ -163,6 +143,194 @@ class ProtobufPlugin implements Plugin<Project> {
+          task.description = "Extracts proto files/dependencies specified by 'protobuf' configuration"
+          task.destDir.set(project.file("${project.buildDir}/extracted-protos/${protoSourceSet.name}"))
+          task.inputFiles.from(protobufConf)
+        } as Action<ProtobufExtract>)
+        protoSourceSet.proto.srcDir(extractProtosTask)
+
+        //
+        // extract include protos task
+        //
+        // In Java projects, the compileClasspath of the 'test' sourceSet includes all the
+        // 'resources' of the output of 'main', in which the source protos are placed.  This is
+        // nicer than the ad-hoc solution that Android has, because it works for any extended
+        // configuration, not just 'testCompile'.
+        String extractIncludeProtosTaskName = protoSourceSet.getTaskName("extractInclude", "proto")
+        TaskProvider<ProtobufExtract> extractIncludeProtosTask = project.tasks.register(extractIncludeProtosTaskName, ProtobufExtract)
+        extractIncludeProtosTask.configure({ ProtobufExtract task ->
+          task.description = "Extracts proto files from compile dependencies for includes"
+          task.destDir.set(project.file("${project.buildDir}/extracted-include-protos/${protoSourceSet.name}"))
+          task.inputFiles.from(compileProtoPathConf)
+        } as Action<ProtobufExtract>)
+        protoSourceSet.includeProtoDirs.srcDir(extractIncludeProtosTask)
+
+        //
+        // generate proto task
+        //
+        String generateProtoTaskName = protoSourceSet.getTaskName("generate", "proto")
+        TaskProvider<GenerateProtoTask> generateProtoTask = project.tasks.register(generateProtoTaskName, GenerateProtoTask)
+        generateProtoTask.configure({ GenerateProtoTask task ->
+          task.description = "Compiles Proto source for '${protoSourceSet.name}'"
+          task.outputBaseDir = project.providers.provider {
+            "${this.protobufExtension.generatedFilesBaseDir}/${protoSourceSet.name}".toString()
+          }
+          task.addSourceDirs(protoSourceSet.proto)
+          task.addIncludeDir(protoSourceSet.proto.sourceDirectories)
+          task.addIncludeDir(protoSourceSet.includeProtoDirs.sourceDirectories)
+
+          if (Utils.isTest(protoSourceSet.name)) {
+            // TL; DR: Make protos in 'test' sourceSet able to import protos from the 'main'
+            // sourceSet.  Sub-configurations, e.g., 'testCompile' that extends 'compile', don't
+            // depend on the their super configurations. As a result, 'testCompile' doesn't depend on
+            // 'compile' and it cannot get the proto files from 'main' sourceSet through the
+            // configuration.
+            ProtoSourceSet mainProtoSourceSet = protobufExtension.sourceSets.getByName(SourceSet.MAIN_SOURCE_SET_NAME)
+            task.addIncludeDir(mainProtoSourceSet.proto.sourceDirectories)
+            task.addIncludeDir(mainProtoSourceSet.includeProtoDirs.sourceDirectories)
```

Comment:
`compileProtoPath` includes `implementation` and `compileOnly`. Should be fine to depend on `compileProtoPath` instead of `compileClasspath`.

![image](https://user-images.githubusercontent.com/18681794/198905058-6ff72fef-0ff0-4b38-b083-c121560c59d7.png)